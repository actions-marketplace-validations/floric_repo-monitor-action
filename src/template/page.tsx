import * as React from "react";
import * as ReactDOM from "react-dom/server";
import { CanvasRenderService } from "chartjs-node-canvas";
import * as dayjs from "dayjs";
import * as localizedFormat from "dayjs/plugin/localizedFormat";
import {
  ReleaseYear,
  MetricsData,
  ReleaseMap,
  Config,
  MetricsContext,
  MetricConfig,
} from "../model";
import { Footer } from "./components/Footer";
import { Header } from "./components/Header";
import { Releases } from "./components/Releases";
import { Metrics } from "./components/Metrics";
import { Page } from "./components/layout/Page";
import { generateLineChart } from "./visualizations/line";
import { importConfig } from "./config/importer";
import { getContent } from "../io/github";

dayjs.extend(localizedFormat);

export const generatePage = async (
  releases: ReleaseYear,
  context: MetricsContext
) => {
  const releasesMap = new Map<string, number>();
  releases.releases
    .sort((a, b) => b.timestamp - a.timestamp)
    .forEach((r, i) => releasesMap.set(r.id, releases.releases.length - i));
  const config = await importConfig(context);
  const data = await getAllData(config, context);
  const graphics = await generateGraphics(data, config, releasesMap);
  return `<!DOCTYPE html>
<html>
  <head>
    <title>${context.owner}/${context.repo} | Metrics</title>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/tailwindcss@1.6.0/dist/tailwind.min.css"
      integrity="sha256-Y4vGjLmrpriLD3X1h1YdyzE2icdiBsJHBXORYXlyDwM="
      crossorigin="anonymous"
    />
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.14.0/css/all.css"
      integrity="sha256-HmKKK3VimMDCOGPTx1mp/5Iaip6BWMZy5HMhLc+4o9E="
      crossorigin="anonymous"
    />
  </head>
  <body>
    ${ReactDOM.renderToStaticMarkup(
      <Page>
        <Header year={releases} repo={context.repo} owner={context.owner} />
        <Releases year={releases} releasesMap={releasesMap} />
        <Metrics config={config} graphics={graphics} />
        <Footer />
      </Page>
    )}
  </body>
</html>`;
};

export const generateGraphics = async (
  data: Array<MetricsData>,
  config: Config,
  releasesMap: ReleaseMap
): Promise<
  Map<string, { img: string; config: MetricConfig; data: MetricsData }>
> => {
  const allMetrics = await Promise.all(
    data.map(async (data) => {
      const configForKey = config.metrics[data.key];
      const img = await renderImage(data, releasesMap);
      return {
        data,
        img,
        config: configForKey,
      };
    })
  );

  const resultMap = new Map();
  allMetrics.forEach((m) => {
    resultMap.set(m.data.key, m);
  });
  return resultMap;
};

const renderImage = async (data: MetricsData, releasesMap: ReleaseMap) => {
  const canvasRenderService = new CanvasRenderService(300, 300);
  const buffer = await canvasRenderService.renderToBuffer(
    generateLineChart(data, releasesMap)
  );
  return buffer.toString("base64");
};

const getAllData = async (config: Config, context: MetricsContext) => {
  const keysToFetch = new Set(
    Object.values(config.groups)
      .map((g) => g.metrics)
      .reduce((a, b) => [...a, ...b], [])
  );
  const rawData = await Promise.all(
    Array.from(keysToFetch).map((key) =>
      getContent(
        context,
        `data/values/${new Date().getUTCFullYear()}/${key}.json`
      )
    )
  );
  const data = rawData
    .filter((n) => !!n.existingSha)
    .map((n) => JSON.parse(n.serializedData));
  return data;
};
