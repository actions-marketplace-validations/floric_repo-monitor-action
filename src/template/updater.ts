import * as core from "@actions/core";
import { getContent, createOrUpdateContent } from "../io/github";
import { MetricsContext, ReleaseYear, MetricsData } from "../model";
import { generatePage } from "./page";

export async function updateTemplate(
  context: MetricsContext,
  releases: ReleaseYear,
  data: Array<MetricsData>
) {
  const { existingSha } = await getContent(context, "index.html");

  const template = await generatePage(releases, data, context);

  core.info(`Generated page successfully`);

  await createOrUpdateContent(context, "index.html", template, existingSha);

  core.info(`Updated page successfully`);
}
