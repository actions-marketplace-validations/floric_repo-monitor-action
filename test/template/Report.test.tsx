import * as React from "react";
import * as renderer from "react-test-renderer";
import { Report } from "../../src/template/Report";

describe("Report", () => {
  it("Renders correctly", () => {
    const tree = renderer.create(
      <Report
        config={{ groups: {}, metrics: {} }}
        context={{
          branch: "master",
          releaseId: "rel-a",
          owner: "floric",
          repo: "repo",
          token: "token",
        }}
        graphics={new Map()}
        releases={{ releases: [], year: 2020 }}
        releasesMap={new Map()}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
