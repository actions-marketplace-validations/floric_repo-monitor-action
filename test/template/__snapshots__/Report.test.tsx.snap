// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report renders correctly 1`] = `
<div
  className="bg-gray-100 text-gray-900 flex p-4 justify-center"
>
  <div
    className="w-full max-w-6xl"
  >
    <h1
      className="text-6xl mb-2"
    >
      <span
        className="font-thin"
      >
        floric
         / 
        repo
         |
      </span>
       
      2020
    </h1>
    <div>
      <h2
        className="text-xl mb-2 mt-8 font-bold"
      >
        Releases
      </h2>
      <table
        className="table-auto w-full text-left"
      >
        <thead>
          <tr
            className="bg-gray-800 text-gray-100"
          >
            <th
              className="px-4 py-2"
            >
              #
            </th>
            <th
              className="px-4 py-2"
            >
              Date
            </th>
            <th
              className="px-4 py-2"
            >
              Commit
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
    <div>
      <h2
        className="text-xl mb-2 mt-8 font-bold"
      >
        Metrics
      </h2>
    </div>
    <div
      className="m-8"
    >
      <p
        className="text-center"
      >
        Last generated by
         
        <a
          className="font-bold"
          href="https://github.com/floric/repo-monitor-action"
        >
          <i
            className="fab fa-github"
          />
           floric/repo-monitor-action
        </a>
         
        on 
        lll
        .
      </p>
    </div>
  </div>
</div>
`;
